ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,536 @@
+            &mut verifier_left_batch,
+            &mut record_counter,
+            ctx,
+            &lagrange_table,
+            Self::polynomials_from_blocks::<R, _>(uv_tuple_block),
+        );
+
+        // recursively generate proofs
+        // until there are less than `R::USIZE * (R::USIZE-1)` many (u,v) field elements
+        // and then do one more iteration
+        // This ensures that after the last iteration,
+        // there are at most R::USIZE-1 many real (u,v) field elements
+        // Therefore at least the last field element is a `F:ZERO` filler that can be replaced
+        // Therefore we have space to include the masks during the final proof
+        //
+        // notice that uv_store.len() will always be at least 1
+        // further, polynomial uv_store.uv[0].0 and .1 will always have
+        // R::USIZE many points since they are filled with `F::ZERO`.
+        loop {
+            let stop = uv_store.len() < R::USIZE * (R::USIZE - 1);
+
+            // generate next proof
+            uv_store = Self::compute_next_proof(
+                &mut prover_left_batch,
+                &mut verifier_left_batch,
+                &mut record_counter,
+                ctx,
+                &lagrange_table,
+                uv_store.iter(),
+            );
+
+            if stop {
+                break;
+            }
+        }
+
+        // generate masks
+        // The right masks are used during the proof generation.
+        // The left masks are later used during the verification.
+        //
+        // Notice that this helper party's left masks are the helper party on the left's right masks
+        // which it uses for its proof generation
+        // Therefore, when this helper party plays the role of a verifier
+        // (for the proof from the party on the left)
+        // it will need to use the masks.
+        let (left_p_mask, right_p_mask): (F, F) =
+            ctx.prss().generate_fields(RecordId::from(record_counter));
+        record_counter += 1;
+        let (left_q_mask, right_q_mask): (F, F) =
+            ctx.prss().generate_fields(RecordId::from(record_counter));
+        record_counter += 1;
+
+        // set masks
+        uv_store.set_masks(right_p_mask, right_q_mask);
+
+        // generate last proof
+        _ = Self::compute_next_proof(
+            &mut prover_left_batch,
+            &mut verifier_left_batch,
+            &mut record_counter,
+            ctx,
+            &lagrange_table,
+            uv_store.iter(),
+        );
+
+        // output proofs
+        // as well as the left masks
+        // which are later needed during the verification
+        (
+            ProofBatch {
+                proofs: prover_left_batch,
+            },
+            ProofBatch {
+                proofs: verifier_left_batch,
+            },
+            (left_p_mask, left_q_mask),
+        )
+    }
+
+    /// This function is a helper function that computes the next proof from an iterator
```

Comment:
two functions would be ideally functions of  ZeroKnowledgeProof, i.e. compute_next_proof and share_via_prss. However, they both use Prss and @benjaminsavage intentionally did not want to us PRSS in the malicious security module since it would make that module more involved. 

For the other one, I could convert UVTupleBlock into a struct and define polynomials_from_blocks as an associated function.